src/language/rule/abstractRule.ts

```diff
@@ -57,11 +57,17 @@ export abstract class AbstractRule implements IRule {
 
     public abstract apply(sourceFile: ts.SourceFile, languageService: ts.LanguageService): RuleFailure[];
 
-    public applyWithWalker(walker: IWalker): RuleFailure[] {
+    protected applyWithWalker(walker: IWalker): RuleFailure[] {
         walker.walk(walker.getSourceFile());
         return this.filterFailures(walker.getFailures());
     }
 
+    protected applyWithFunction<T>(sourceFile: ts.SourceFile, options: T, walkFn: (ctx: WalkContext<T>) => void) {
```

Comment:
maybe `options` should be a third optional param? Since it's sometimes `undefined`?